mlflow/tensorflow/__init__.py

```diff
@@ -887,6 +968,110 @@ def _log_early_stop_callback_metrics(callback, history, metrics_logger):
         if metric_key is not None:
             metrics_logger.record_metrics(restored_metrics, stopped_epoch + 1)
 
+    def _log_keras_model_with_warning(history, args):
+        try:
+            if log_models:
+                import types
+
+                def _get_input_data_slice():
+                    input_training_data_original = args[0]
+                    input_example_slice = None
+
+                    def _extract_from_sequence_or_generator():
+                        input_training_data = input_training_data_original
+                        if isinstance(input_training_data_original, types.GeneratorType):
+                            input_training_data = next(input_training_data_original)[0]
```

Comment:
Does calling `next()` on the data affect training execution at all? I.e. by consuming part of this iterator, are we preventing the model from consuming it during training? If so, we should make a copy of the object before mutating it and add test coverage to make sure we aren't mutating the training dataset.

Reply:
As discussed with you offline, `GeneratorType` is going to be a separate PR that I will raise soon.